WordPress/gutenberg
# Add new preferences persistence API, and save editor preferences in user meta

Related - https://github.com/WordPress/gutenberg/issues/31965
Also related - https://github.com/WordPress/gutenberg/pull/39672#discussion_r832967226

Fixes #15105
Fixes #34405

## What?
This PR explores moving away from the current `@wordpress/data` way of persisting store data to local storage.

It implements a new API in the preferences package for configuring how the store state is persisted.

It adds a new package that persists preferences to the WordPress database, as part of the user's meta.

And then finally this is all configured using an inline script.

## Why?
A good reason is that this will finally allow https://github.com/WordPress/gutenberg/issues/15105 to be closed, which is a long standing problem for users.

There are also some technical advantages that this will unlock:
- deprecating the old persistence system in `@wordpress/data`
- deprecating `registerStore`in `@wordpress/data` in favor of `register`

## How?
Adds a new `setPersistenceLayer` action to the preferences package. A higher order reducer is then used to keep state in sync with the persistence layer in the preferences store.

There's also a new `preferences-persistence` package (based on https://github.com/WordPress/gutenberg/pull/19177).  This saves data to user meta via the REST API primarily. It also saves data to local storage as a back in case the user goes offline or a request is interrupted.

Finally, I've moved all the migration code over from the old package to this new package, any old preferences should be retained.

## Testing Instructions
1. Open up two different browsers side by side.
2. Login to your test environment in both browsers using the same login and edit a post, site or widgets
3. Change some settings in one browser (i.e. top toolbar)
4. Reload the other browser and you should see the change reflected.
